util/hot_backup.cc

```diff
@@ -0,0 +1,654 @@
+        }   // if
+    }   // if
+
+    return(good);
+
+}   // DBImpl::PrepareDirectories
+
+
+/**
+ * Goal is to ensure the write buffer has not been sitting
+ * around for hours or days without a backup.  Not looking
+ * for perfect, up to the second backup of a constant input
+ * stream.
+ */
+bool
+DBImpl::PurgeWriteBuffer()
+{
+    bool good(true);
+    Status s;
+
+    // existing "test"
+    s=CompactMemTableSynchronous();
+    good=s.ok();
+
+    if (!good)
+        Log(GetLogger(), "HotBackup failed in PurgeWriteBuffer");
+
+    return(good);
+
+}   // DBImpl::PurgeWriteBuffer
+
+
+/**
+ * Need an independent MANIFEST and CURRENT file within
+ * hot backup directory.  This creates them.
+ */
+bool
+DBImpl::WriteBackupManifest()
+{
+    bool good(true);
+    Version * version(NULL);
+    uint64_t manifest_num;
+    std::string manifest_string;
+    SequenceNumber sequence;
+    VersionEdit edit;
+    int level;
+    Options local_options;
+    Status status;
+
+    // perform all non-disk activity quickly while mutex held
+    {
+        MutexLock l(&mutex_);
+//        InternalKeyComparator const icmp(BytewiseComparator());
+        InternalKeyComparator const icmp(internal_comparator_.user_comparator());
+
+        version=versions_->current();
+        version->Ref();               // must have mutex for Ref/Unref
+
+        // reserve a file number for manifest, get a recent
+        //  sequence number for "highest" (exact not needed)
+        manifest_num=versions_->NewFileNumber();
+        sequence=versions_->LastSequence();
+
+        // lines taken from repair.cc
+        edit.SetComparatorName(icmp.user_comparator()->Name());
+        edit.SetLogNumber(0);
+        edit.SetNextFile(versions_->NewFileNumber());
+        edit.SetLastSequence(sequence);
+    }   // mutex
+
+    // copy list of files from Version to VersionEdit
+    for (level=0; level<config::kNumLevels; ++level)
+    {
+        const Version::FileMetaDataVector_t level_files(version->GetFileList(level));
+        Version::FileMetaDataVector_t::const_iterator it;
+
+        for (it=level_files.begin(); it!=level_files.end(); ++it)
+        {
+            edit.AddFile2(level, (*it)->number, (*it)->file_size,
+                          (*it)->smallest, (*it)->largest,
+                          (*it)->exp_write_low, (*it)->exp_write_high, (*it)->exp_explicit_high);
+        }   // for
+    }   // for
+
+    // create Manifest and CURRENT in backup dir
+    local_options=options_;
+    SetBackupPaths(local_options, 0);
+    manifest_string=DescriptorFileName(local_options.tiered_fast_prefix, manifest_num);
+
+    WritableFile* file;
+    status = env_->NewWritableFile(manifest_string, &file, 4096);
+    if (status.ok())
+    {
+        log::Writer log(file);
+        std::string record;
+        edit.EncodeTo(&record);
+        status = log.AddRecord(record);
+
+        if (status.ok())
+            status = file->Close();
+
+        delete file;
```

Comment:
C++11 added std::unique_ptr<>, which supersedes std::auto_ptr<>, and adds
some additional goodness (such as properly handling arrays).

I tried numerous times to move our toolset forward so that I could start
using unique_ptr<>, but to no avail. Yes, you're stuck with auto_ptr<> for
now.

On Sun, Sep 4, 2016 at 3:04 PM, Matthew Von-Maszewski <
notifications@github.com> wrote:

> In util/hot_backup.cc
> <https://github.com/basho/leveldb/pull/202#discussion_r77457744>:
>
> > +    SetBackupPaths(local_options, 0);
> > +    manifest_string=DescriptorFileName(local_options.tiered_fast_prefix, manifest_num);
> > +
> > +    WritableFile* file;
> > +    status = env_->NewWritableFile(manifest_string, &file, 4096);
> > +    if (status.ok())
> > +    {
> > +        log::Writer log(file);
> > +        std::string record;
> > +        edit.EncodeTo(&record);
> > +        status = log.AddRecord(record);
> > +
> > +        if (status.ok())
> > +            status = file->Close();
> > +
> > +        delete file;
>
> I read somewhere that std::auto_ptr<> is being deprecated. Do you happen
> to know its replacement?
>
> That said, I will use the auto_ptr since our build environments are way,
> way behind the standards. Question is a personal curiosity.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/basho/leveldb/pull/202/files/9b58e117f4249df583516a6ca3539826525b4b93#r77457744>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMnddNN4F3QY4ioeg5fckfF_OVrgmGqKks5qmyRDgaJpZM4JcAvk>
> .
>